src/server/pfs/server/driver_new_storage.go

```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
 	return nil
 }
 
-func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes []string) (retErr error) {
+// compact is the entrypoint to compaction.
+func (d *driver) compact(ctx context.Context, master *work.Master, outputPath string, inputPrefixes []string) error {
+	// resolve prefixes into paths
+	inputPaths := []string{}
+	for _, inputPrefix := range inputPrefixes {
+		err := d.storage.WalkFileSets(ctx, inputPrefix, func(inputPath string) error {
+			inputPaths = append(inputPaths, inputPath)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	if d.env.StorageCompactionMaxFanIn < 2 {
+		panic("StorageCompactionMaxFanIn cannot be < 2")
+	}
+	return d.compactIter(ctx, compactParams{
+		master:     master,
+		inputPaths: inputPaths,
+		outputPath: outputPath,
+		maxFanIn:   d.env.StorageCompactionMaxFanIn,
+	})
+}
+
+type compactParams struct {
+	master     *work.Master
+	outputPath string
+	inputPaths []string
+	maxFanIn   int
+}
+
+// compactIter is one level of compaction.  It will only perform compaction
+// if len(inputPaths) < params.maxFanIn otherwise it will split inputPaths recursively.
+func (d *driver) compactIter(ctx context.Context, params compactParams) (retErr error) {
+	scratch := path.Join(tmpPrefix, uuid.NewWithoutDashes())
+	defer func() {
+		if err := d.storage.Delete(ctx, scratch); retErr == nil {
+			retErr = err
+		}
+	}()
+	// exceeds maxFanIn need to branch out
+	inputPaths := params.inputPaths
+	if len(inputPaths) > params.maxFanIn {
```

Comment:
Something that I think could be beneficial from a readability standpoint would be to structure this how most recursive functions are where the base case is checked first, then the recursive step is handled after the check. So, you would basically invert what you have now, so the check would be less than or equal to fan in, then run sharded compact. If not, then kick off the recursion and run sharded compact on the results. Just an idea, what are your thoughts?

Reply:
Yeah that could work.  I think it prevents the reassign to inputPaths, which is cleaner.  It also lets us skip the scratch space for the base case.  I'll give it a try.